File: web/test/similar.test.ts
import { expect, test } from "bun:test";
import { similarTo } from "../public/lib/catalog/similar.js";

const t = (setId: string, genres: string[], extra: Record<string, unknown> = {}) => ({ setId, genres, ...extra });

test("franchise first, then unwatched, then genres shared, then popularity", () => {
  const dune = t("dune2", ["Sci-Fi", "Adventure"], { collectionId: 7 });
  const picks = similarTo(dune, [
    dune,
    t("dune1", ["Drama"], { collectionId: 7 }),
    t("arrival", ["Sci-Fi"], { popularity: 50 }),
    t("interstellar", ["Sci-Fi", "Adventure"], { popularity: 10 }),
    t("seen", ["Sci-Fi", "Adventure"], { popularity: 99 }),
    t("romcom", ["Romance"]),
  ], (item) => item.setId === "seen");
  expect(picks.map((item) => item.setId)).toEqual(["dune1", "interstellar", "arrival", "seen"]);
});

test("a title with nothing in common with anything has no similar titles", () => {
  expect(similarTo(t("a", []), [t("b", ["Drama"])])).toEqual([]);
});

File: android/feature/catalog/src/test/kotlin/SimilarTest.kt
package catalog

import model.Kind
import model.MediaSet
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/** Mirrors `web/test/similar.test.ts`, case for case. */
class SimilarTest {
    private fun title(
        setId: String,
        genres: List<String> = emptyList(),
        collectionId: Long? = null,
        popularity: Double? = null,
    ): MediaSet =
        MediaSet(
            setId = setId, kind = Kind.MOVIE, title = setId, show = null, chapter = null, path = null,
            season = null, episodeFirst = null, episodeLast = null, year = null, durationSecs = null,
            posterPath = null, totalBytes = 0, genres = genres, collectionId = collectionId, popularity = popularity,
        )

    @Test
    fun franchiseFirstThenUnwatchedThenGenresSharedThenPopularity() {
        val dune = title("dune2", listOf("Sci-Fi", "Adventure"), collectionId = 7)
        val picks =
            similarTo(
                dune,
                listOf(
                    dune,
                    title("dune1", listOf("Drama"), collectionId = 7),
                    title("arrival", listOf("Sci-Fi"), popularity = 50.0),
                    title("interstellar", listOf("Sci-Fi", "Adventure"), popularity = 10.0),
                    title("seen", listOf("Sci-Fi", "Adventure"), popularity = 99.0),
                    title("romcom", listOf("Romance")),
                ),
                seen = { it.setId == "seen" },
            )
        assertEquals(listOf("dune1", "interstellar", "arrival", "seen"), picks.map(MediaSet::setId))
    }

    @Test
    fun aTitleWithNothingInCommonWithAnythingHasNoSimilarTitles() {
        assertTrue(similarTo(title("a"), listOf(title("b", listOf("Drama")))).isEmpty())
    }
}
